types/context_cache.go

```diff
@@ -37,73 +31,53 @@ func (c *ContextMemCache) UpsertDeferredSends(moduleAccount string, amount Coins
-	ok := c.deferredWithdrawals.SafeSub(moduleAccount, amount)
-	if !ok {
-		c.deferredSends.UpsertMapping(moduleAccount, amount)
-	}
+	c.deferredSends.UpsertMapping(moduleAccount, amount)
 	return nil
 }
 
-func (c *ContextMemCache) SafeSubDeferredSends(moduleAccount string, amount Coins) bool {
+// This will perform a saturating sub against the deferred module balance atomically. This means that it will subtract any balances that it is able to, and then call the passed in `remainder handler` to ensure the remaining balance can be safely subtracted as well. If this remainderHandler returns an error, we will revert the saturating sub to ensure atomicity of the subtraction across the multiple balances
+func (c *ContextMemCache) SaturatingSubDeferredSends(moduleAccount string, amount Coins, remainderHandler func(amount Coins) error) error {
 	if !amount.IsValid() {
 		panic(sdkerrors.Wrap(sdkerrors.ErrInvalidCoins, amount.String()))
 	}
-
-	return c.deferredSends.SafeSub(moduleAccount, amount)
-}
-
-func (c *ContextMemCache) RangeOnDeferredSendsAndDelete(apply func(recipient string, amount Coins)) {
-	c.deferredSends.RangeOnMapping(apply)
-}
-
-func (c *ContextMemCache) UpsertDeferredWithdrawals(moduleAccount string, amount Coins) error {
-	if !amount.IsValid() {
-		return sdkerrors.Wrap(sdkerrors.ErrInvalidCoins, amount.String())
-	}
-
-	// Separate locks needed for all mappings - atmoic transaction needed
 	c.deferredBankOpsLock.Lock()
 	defer c.deferredBankOpsLock.Unlock()
 
-	// If there's already a pending deposit then subtract it from that amount first
-	// or else add it to the deferredWithdrawals mapping
-	ok := c.deferredSends.SafeSub(moduleAccount, amount)
+	originalBalance, ok := c.deferredSends.Get(moduleAccount)
 	if !ok {
-		c.deferredWithdrawals.UpsertMapping(moduleAccount, amount)
+		// run the remainder handler on the full amount
+		err := remainderHandler(amount)
+		// no need for revert because we dont even attempt a saturating sub, bubble up error if applicable
+		return err
 	}
-	return nil
+	// found a balance, try to perform the logic
+	remainder := c.deferredSends.SaturatingSub(moduleAccount, amount)
+	if remainder.IsZero() {
+		// no remainder, return now
+		return nil
+	}
+	err := remainderHandler(remainder)
+	if err != nil {
+		// revert the map subtraction and the bubble up error
+		c.deferredSends.Set(moduleAccount, originalBalance)
+	}
+	return err
 }
 
-// This inserts or updates an entry for a module account for a deferred withdrawal.
-// This should be performed AFTER checking for a sufficient balance in the underlying bank balances for that module account.
-// Additionally, this does not attempt to do any safe subtraction from pending sends, so if that behavior is preferred, it will need to be checked separately.
-func (c *ContextMemCache) UpsertDeferredWithdrawalsNoSafeSub(moduleAccount string, amount Coins) error {
+func (c *ContextMemCache) SafeSubDeferredSends(moduleAccount string, amount Coins) bool {
```

Comment:
from the PR description, it seems like this is not meant to be called publicly. Can we change this to start w/ lower case so it's not exported?

Reply:
do we still want to export it so that the unit tests for it are still able to run?

Reply:
do we need mutex lock on this? looks like we're doing a write

Reply:
It should still work if called in any files ending in test.go i believe

Reply:
I can add a mutex here, but the original implementation didn't have one (but probably safer to lock)